File: tools/lzo1x.py
"""
lzo1x.py -- LZO1X 解压（miniLZO 解码器的 Python 直译）。

为什么要它：RA2/YR 地图的 [IsoMapPack5] / [OverlayPack] 段是
**分块 LZO + base64**（modenc 的 IsoMapPack5 页面写明用 lzo1x_decompress）。
不解压就看不到任何一个瓦片。

直译自 lib/lzo/lzo1x_decompress_safe.c（Linux 内核里的 miniLZO 移植版），
但去掉了两样 RA2 用不到的东西：
  1. bitstream_version（LZO 2.05+ 才有的零游走码），RA2 是 LZO 1.x/2.02 时代，恒为 0
  2. 各种 unaligned 快速路径 —— 只留语义正确的慢路径

状态机说明（内核版用 state 表示，原版用 goto 位置表示）：
  state == 0  上一次是"匹配"，本轮 token<16 => 字面量跑
  state == 4  上一次是"字面量跑"，token<16 => 3 字节 M1 匹配
  state 其它  token<16 => 2 字节匹配 + next 个尾随字面量
"""

from __future__ import annotations

M1_MAX_OFFSET = 0x0400
M2_MAX_OFFSET = 0x0800
M3_MAX_OFFSET = 0x4000
M4_MAX_OFFSET = 0xBFFF


class LzoError(Exception):
    pass


def decompress(src: bytes, out_max: int) -> bytes:
    """把 src（一个 LZO1X 块）解压，最多写 out_max 字节。"""
    out = bytearray()
    ip = 0
    ip_end = len(src)

    def need_ip(n: int):
        if ip_end - ip < n:
            raise LzoError("input overrun")

    def need_op(n: int):
        if out_max - len(out) < n:
            raise LzoError("output overrun")

    state = 0
    t = 0
    next_ = 0

    if ip_end >= 1 and src[0] > 17:
        t = src[ip] - 17
        ip += 1
        if t < 4:
            next_ = t
            # 直接跳到 match_next 的语义
            need_ip(t + 1)
            need_op(t)
            out += src[ip:ip + t]
            ip += t
        else:
            need_ip(t + 1)
            need_op(t)
            out += src[ip:ip + t]
            ip += t
            state = 4

    while ip < ip_end:
        t = src[ip]
        ip += 1
        if t < 16:
            if state == 0:
                if t == 0:
                    ip_last = ip
                    while src[ip] == 0:
                        ip += 1
                        need_ip(1)
                    offset = (ip - ip_last) * 255
                    t += offset + 15 + src[ip]
                    ip += 1
                t += 3
                # copy_literal_run
                need_ip(t + 1)
                need_op(t)
                out += src[ip:ip + t]
                ip += t
                state = 4
                continue
            elif state != 4:
                next_ = t & 3
                m_pos = len(out) - 1 - (t >> 2) - (src[ip] << 2)
                ip += 1
                if m_pos < 0:
                    raise LzoError("lookbehind overrun")
                need_op(2)
                out += out[m_pos:m_pos + 2]
            else:
                next_ = t & 3
                m_pos = len(out) - (1 + M2_MAX_OFFSET) - (t >> 2) - (src[ip] << 2)
                ip += 1
                t = 3
        elif t >= 64:
            next_ = t & 3
            m_pos = len(out) - 1 - ((t >> 2) & 7) - (src[ip] << 3)
            ip += 1
            t = ((t >> 5) - 1) + 2
        elif t >= 32:
            t = (t & 31) + 2
            if t == 2:
                ip_last = ip
                while src[ip] == 0:
                    ip += 1
                    need_ip(1)
                offset = (ip - ip_last) * 255
                t += offset + 31 + src[ip]
                ip += 1
                need_ip(2)
            next_ = src[ip] | (src[ip + 1] << 8)      # le16
            ip += 2
            m_pos = len(out) - 1 - (next_ >> 2)
            next_ &= 3
        else:
            need_ip(2)
            next_ = src[ip] | (src[ip + 1] << 8)      # le16
            m_pos = len(out) - ((t & 8) << 11)
            t = (t & 7) + 2
            if t == 2:
                ip_last = ip
                while src[ip] == 0:
                    ip += 1
                    need_ip(1)
                offset = (ip - ip_last) * 255
                t += offset + 7 + src[ip]
                ip += 1
                need_ip(2)
                next_ = src[ip] | (src[ip + 1] << 8)
            ip += 2
            m_pos -= next_ >> 2
            next_ &= 3
            if m_pos == len(out):
                break                                  # eof_found
            m_pos -= 0x4000

        if m_pos < 0:
            raise LzoError("lookbehind overrun")
        need_op(t)
        # 逐字节拷贝（匹配允许重叠，不能用切片一次性拷）
        for _ in range(t):
            out.append(out[m_pos])
            m_pos += 1

        # match_next
        state = next_
        t = next_
        need_ip(t + 1)
        need_op(t)
        if t:
            out += src[ip:ip + t]
            ip += t

    return bytes(out)


def decompress_chunks(blob: bytes) -> bytes:
    """RA2 的分块格式：重复 [u16 输入长度][u16 输出长度][输入数据]。"""
    out = bytearray()
    p = 0
    n = len(blob)
    while p + 4 <= n:
        in_sz = blob[p] | (blob[p + 1] << 8)
        out_sz = blob[p + 2] | (blob[p + 3] << 8)
        p += 4
        if in_sz == 0 or out_sz == 0:
            break
        if p + in_sz > n:
            raise LzoError("块输入越界: p=%d in_sz=%d n=%d" % (p, in_sz, n))
        chunk = decompress(blob[p:p + in_sz], out_sz)
        if len(chunk) != out_sz:
            raise LzoError("块解压长度不符: 得到 %d，期望 %d" % (len(chunk), out_sz))
        out += chunk
        p += in_sz
    return bytes(out)

File: src/io/Lzo1x.cpp
// Lzo1x.cpp -- LZO1X 解压实现。规则说明见 Lzo1x.h。

#include "io/Lzo1x.h"

namespace ra2 {
namespace {

constexpr size_t kM2MaxOffset = 0x0800;

inline uint16_t Rd16(const uint8_t* p) {
    return static_cast<uint16_t>(p[0] | (p[1] << 8));   // 小端
}

}  // namespace

int Lzo1x_Decompress(const uint8_t* in, size_t in_len,
                     uint8_t* out, size_t out_cap, size_t* out_len) {
    if (in == nullptr || out == nullptr || out_len == nullptr || in_len < 3) {
        return kLzoInputOverrun;
    }

    size_t ip = 0;
    size_t op = 0;
    const size_t ip_end = in_len;

    // state: 0 = 刚消费完"尾随字面量"（下一轮 token<16 是字面量跑）
    //        4 = 刚消费完"字面量跑"（下一轮 token<16 是 3 字节 M1 匹配）
    //      其它 = 刚消费完匹配
    size_t state = 0;
    size_t t = 0;
    size_t next = 0;
    size_t m_pos = 0;

    auto need_ip = [&](size_t n) { return ip_end - ip >= n; };
    auto need_op = [&](size_t n) { return out_cap - op >= n; };

    // 开头特殊情形：第一个字节 > 17，说明起始就是一段字面量。
    if (in[0] > 17) {
        t = in[ip++] - 17;
        if (t < 4) {
            // 起手 1..3 个字面量，且没有前置匹配。
            if (!need_ip(t) || !need_op(t)) {
                *out_len = op;
                return kLzoInputOverrun;
            }
            for (size_t i = 0; i < t; ++i) out[op++] = in[ip++];
            // 与 match_next 一致：state 记成 t，下一轮按"匹配后"处理。
            state = t;
            next = t;
            t = next;
            // 直接进入主循环的 match_next 逻辑：这里没有匹配，只把尾随字面量吃掉。
            // 已经拷贝过了，所以跳过一次。
            goto after_match_next;
        }
        if (!need_ip(t) || !need_op(t)) {
            *out_len = op;
            return kLzoInputOverrun;
        }
        for (size_t i = 0; i < t; ++i) out[op++] = in[ip++];
        state = 4;
    }

    while (ip < ip_end) {
        t = in[ip++];
        if (t < 16) {
            if (state == 0) {
                if (t == 0) {
                    // 字面量跑长度扩展：连续 0 字节每段代表 +255。
                    size_t zeros = 0;
                    while (ip < ip_end && in[ip] == 0) {
                        ++zeros;
                        ++ip;
                    }
                    if (!need_ip(1)) {
                        *out_len = op;
                        return kLzoInputOverrun;
                    }
                    t += zeros * 255 + 15 + in[ip++];
                }
                t += 3;
                if (!need_ip(t) || !need_op(t)) {
                    *out_len = op;
                    return (need_ip(t) ? kLzoOutputOverrun : kLzoInputOverrun);
                }
                for (size_t i = 0; i < t; ++i) out[op++] = in[ip++];
                state = 4;
                continue;
            } else if (state != 4) {
                // 2 字节匹配 + 尾随字面量（M1）
                if (!need_ip(1)) {
                    *out_len = op;
                    return kLzoInputOverrun;
                }
                next = t & 3;
                m_pos = op - 1 - (t >> 2) - (static_cast<size_t>(in[ip++]) << 2);
                if (m_pos > op || !need_op(2)) {
                    *out_len = op;
                    return (m_pos > op ? kLzoLookbehindOverrun : kLzoOutputOverrun);
                }
                out[op++] = out[m_pos++];
                out[op++] = out[m_pos++];
            } else {
                // 3 字节匹配（紧跟字面量跑）
                if (!need_ip(1)) {
                    *out_len = op;
                    return kLzoInputOverrun;
                }
                next = t & 3;
                m_pos = op - (1 + kM2MaxOffset) - (t >> 2)
                        - (static_cast<size_t>(in[ip++]) << 2);
                t = 3;
            }
        } else if (t >= 64) {
            if (!need_ip(1)) {
                *out_len = op;
                return kLzoInputOverrun;
            }
            next = t & 3;
            m_pos = op - 1 - ((t >> 2) & 7) - (static_cast<size_t>(in[ip++]) << 3);
            t = ((t >> 5) - 1) + 2;
        } else if (t >= 32) {
            t = (t & 31) + 2;
            if (t == 2) {
                size_t zeros = 0;
                while (ip < ip_end && in[ip] == 0) {
                    ++zeros;
                    ++ip;
                }
                if (!need_ip(1)) {
                    *out_len = op;
                    return kLzoInputOverrun;
                }
                t += zeros * 255 + 31 + in[ip++];
            }
            if (!need_ip(2)) {
                *out_len = op;
                return kLzoInputOverrun;
            }
            next = Rd16(in + ip);
            ip += 2;
            m_pos = op - 1 - (next >> 2);
            next &= 3;
        } else {
            if (!need_ip(2)) {
                *out_len = op;
                return kLzoInputOverrun;
            }
            next = Rd16(in + ip);
            m_pos = op - (static_cast<size_t>(t & 8) << 11);
            t = (t & 7) + 2;
            if (t == 2) {
                size_t zeros = 0;
                while (ip < ip_end && in[ip] == 0) {
                    ++zeros;
                    ++ip;
                }
                if (!need_ip(1)) {
                    *out_len = op;
                    return kLzoInputOverrun;
                }
                t += zeros * 255 + 7 + in[ip++];
                if (!need_ip(2)) {
                    *out_len = op;
                    return kLzoInputOverrun;
                }
                next = Rd16(in + ip);
            }
            ip += 2;
            m_pos -= next >> 2;
            next &= 3;
            if (m_pos == op) {
                // M4 偏移为 0 == 流结束标记
                *out_len = op;
                return kLzoOk;
            }
            m_pos -= 0x4000;
        }

        if (m_pos > op || !need_op(t)) {
            *out_len = op;
            return (m_pos > op ? kLzoLookbehindOverrun : kLzoOutputOverrun);
        }
        // 逐字节拷贝：匹配区间与输出区间允许重叠，不能用 memcpy。
        for (size_t i = 0; i < t; ++i) out[op++] = out[m_pos++];

        // match_next：匹配后还有 next 个字面量，直接从输入搬。
        state = next;
        t = next;
        if (t > 0) {
            if (!need_ip(t) || !need_op(t)) {
                *out_len = op;
                return (need_ip(t) ? kLzoOutputOverrun : kLzoInputOverrun);
            }
            for (size_t i = 0; i < t; ++i) out[op++] = in[ip++];
        }
    after_match_next:;
    }

    *out_len = op;
    return kLzoOk;
}

bool Lzo1x_Decompress_Chunks(const uint8_t* blob, size_t size,
                             std::vector<uint8_t>* out, const char** err) {
    out->clear();
    size_t p = 0;
    while (p + 4 <= size) {
        const size_t in_sz = Rd16(blob + p);
        const size_t out_sz = Rd16(blob + p + 2);
        p += 4;
        if (in_sz == 0 || out_sz == 0) {
            break;
        }
        if (p + in_sz > size) {
            if (err) *err = "LZO 块输入越界";
            return false;
        }
        const size_t base = out->size();
        out->resize(base + out_sz);
        size_t got = 0;
        const int rc = Lzo1x_Decompress(blob + p, in_sz, out->data() + base,
                                        out_sz, &got);
        if (rc != kLzoOk || got != out_sz) {
            out->resize(base + got);
            if (err) *err = "LZO 块解压长度与声明不符";
            return false;
        }
        p += in_sz;
    }
    return true;
}

}  // namespace ra2
